### ISSUE TITLE
User namespaces on Funtoo/Gentoo

### ISSUE THREAD

**Issue by Calrama** [docker/docker#20548]
Description of problem:

Cannot use user namespaces on Funtoo/Gentoo.

`docker version`:
```
Client:
 Version:      1.10.0
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   590d510
 Built:        
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.0
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   590d510
 Built:        
 OS/Arch:      linux/amd64
```

`docker info`:
```
Containers: 4
 Running: 0
 Paused: 0
 Stopped: 4
Images: 1
Server Version: 1.10.0
Storage Driver: zfs
 Zpool: system
 Zpool Health: ONLINE
 Parent Dataset: system/docker
 Space Used By Parent: 6516736
 Space Available: 110319980032
 Parent Quota: no
 Compression: on
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: host bridge null
Kernel Version: 4.5.0-rc4
Operating System: Linux
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.706 GiB
Name: cylae
ID: --------------------------
WARNING: No cpu cfs quota support
WARNING: No cpu cfs period support
```

`uname -a`:
```
Linux cylae 4.5.0-rc4 #2 SMP PREEMPT Sat Feb 20 22:42:48 CET 2016 x86_64 Intel(R) Core(TM) i5-5200U CPU @ 2.20GHz GenuineIntel GNU/Linux
```

Environment details (AWS, VirtualBox, physical, etc.):
**OS**: Funtoo Linux (current)

Steps to Reproduce:
1. Start docker daemon with option "--userns-remap=default"
2. ```docker pull alpine```
3. ```docker run -it alpine /bin/sh```

Actual Results:

```
Timestamp: 2016-02-21 01:57:13.003810235 +0100 CET
Code: System error

Message: stat /run/lxcfs/controllers/name=openrc/docker: permission denied

Frames:
---
0: setupRootfs
Package: github.com/opencontainers/runc/libcontainer
File: rootfs_linux.go@40
---
1: Init
Package: github.com/opencontainers/runc/libcontainer.(*linuxStandardInit)
File: standard_init_linux.go@57
---
2: StartInitialization
Package: github.com/opencontainers/runc/libcontainer.(*LinuxFactory)
File: factory_linux.go@240
---
3: initializer
Package: github.com/docker/docker/daemon/execdriver/native
File: init.go@35
---
4: Init
Package: github.com/docker/docker/pkg/reexec
File: reexec.go@26
---
5: main
Package: main
File: docker.go@18
---
6: main
Package: runtime
File: proc.go@111
---
7: goexit
Package: runtime
File: asm_amd64.s@1721
docker: Error response from daemon: Container command could not be invoked..
```

Expected Results:

```/bin/sh``` prompt in container

Additional info:

- cgmanager and lxcfs are setup on the host correctly, lxd is also working perfectly.
- The error above shows the cgroup OpenRC puts the docker daemon into (automatically) in its own controller, so maybe the OpenRC controller it not handled by docker?

**Comment by Calrama** [comment]
I can manually bypass this first error by applying `chmod o+rx` on `/run/lxcfs/controllers/name=openrc/docker` from top to bottom. This, however, then leads to the next error:

```
Timestamp: 2016-02-21 12:00:06.341250015 +0100 CET
Code: System error

Message: can't create pivot_root dir , error mkdir /var/lib/docker/165536.165536/zfs/graph/395e2ecc9815703acef008ea43c6311d4614695b19a030c22209031ca0d6bc60/.pivot_root975280611: permission denied

Frames:
---
0: setupRootfs
Package: github.com/opencontainers/runc/libcontainer
File: rootfs_linux.go@69
---
1: Init
Package: github.com/opencontainers/runc/libcontainer.(*linuxStandardInit)
File: standard_init_linux.go@57
---
2: StartInitialization
Package: github.com/opencontainers/runc/libcontainer.(*LinuxFactory)
File: factory_linux.go@240
---
3: initializer
Package: github.com/docker/docker/daemon/execdriver/native
File: init.go@35
---
4: Init
Package: github.com/docker/docker/pkg/reexec
File: reexec.go@26
---
5: main
Package: main
File: docker.go@18
---
6: main
Package: runtime
File: proc.go@111
---
7: goexit
Package: runtime
File: asm_amd64.s@1721
docker: Error response from daemon: Container command could not be invoked..
```

Sadly, no amount of `chmod` on that path seems to be able to solve this.

**Comment by estesp** [comment]
The second error is fixed in the Docker 1.10.1 bug fix update via #20045.

The first error as you found shows that there is no way to have a non-root user mount filesystems on a path that is inaccessible to that user.  This would be true for user namespaced processes inside or outside of Docker and is simply a file permissions issue that lxcfs isn't handling or aware of.  Note that the syscalls that give you that error only need `x` permissions on the path traversed, so `chmod o+x` is sufficient.

**Comment by Calrama** [comment]
Thank you for the explanation, after upgrading to 1.10.2 (and rebooting), all that is still required to get it to run is `sudo chmod o+x /run/lxcfs/controllers`. On that note: If lxcfs is not handling is / aware of it, I would like to know what - from dockers standpoint - is the most favorable long-term solution:
- Make lxcfs aware of the issue and request to have /run/lxcfs/controllers made traversable for anyone (are there any security issues with having the path traversable that you are aware of?)
- Change docker so it does not need the path traversable (I don't know if this is feasible, I'm just bringing it up because lxd does not seem to need it)
- Have docker detect that lxcfs is running and make the path traversable at daemon startup

Currently, I will just have a startup script accompanying lxcfs to make the path traversable.

**Comment by hallyn** [comment]
/run/lxcfs is private to lxcfs - the paths which containers should bind mount are under /var/lib/lxcfs.

**Comment by brauner** [comment]
This should be fixed by https://github.com/lxc/lxcfs/pull/124.

**Comment by Calrama** [comment]
@brauner Thank you for that, WFM.

For any other Gentoo/Funtoo users that cannot wait for the PR to be accepted:

1. Get the PR [as a patch](https://patch-diff.githubusercontent.com/raw/lxc/lxcfs/pull/124.patch) into `/etc/portage/patches/sys-fs/lxcfs/124.patch`
2. Get yourselves a development, user-patches enabled ebuild of lxcfs, e.g. [mine](https://git.ucworks.org/calrama/gestalt/src/master/sys-fs/lxcfs/lxcfs-9999.ebuild)
3. Emerge the ebuild (remember to stop the lxcfs service)
4. Reboot (may not be necessary depending on your setup. For myself, I got ConsoleKit2 `permission denied` errors until I rebooted - just relogging did not work for me).

`docker version`
```
Client:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        
 OS/Arch:      linux/amd64
```

`docker info`
```
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 1.12.0
Storage Driver: zfs
 Zpool: error while getting pool information strconv.ParseUint: parsing "": invalid syntax
 Zpool Health: not available
 Parent Dataset: system/docker
 Space Used By Parent: 11116544
 Space Available: 11792110592
 Parent Quota: no
 Compression: off
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null host bridge overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.6.4-gentoo
Operating System: Linux
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.707 GiB
Name: -----
ID: -----------------------
Docker Root Dir: /var/lib/docker/362144.362144
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
```

`uname -a`
```
Linux cylae 4.6.4-gentoo #2 SMP PREEMPT Sat Jul 16 10:27:08 CEST 2016 x86_64 Intel(R) Core(TM) i5-5200U CPU @ 2.20GHz GenuineIntel GNU/Linux
```

Environment details (AWS, VirtualBox, physical, etc.):
*LXCFS*: Commit https://github.com/lxc/lxcfs/commit/8e365eef40af04483af61d2b393bc4ddb5350c4e with PR https://github.com/lxc/lxcfs/pull/124

**Comment by thaJeztah** [comment]
Looks like this is fixed, or will be soon, so I'll go ahead and close this; thanks all!
